### ISSUE TITLE
Extend space-in-parens to allow different formatting for ES6 arrow functions

### ISSUE THREAD

**Issue by zeg-io** [eslint/eslint#7016]
**What version are you using?**
v3.4.0

**What did you do?**

``` js
conn.sqlManager = new sql.connect(conn.connectionString).then( (newConn) => {
          if (newConn.connected) {
            log.good('done.')
            newConn.isConnected = true
            next()
          } else {
            newConn.isConnected = false
            errorHandler(res, newConn, {
              code: 'NOT_CONNECT',
              message: 'Database connection error'
            })
            res.status(500).json({
              statusCode: 'ERROR',
              message: 'Database connection error.'
            })
            next()
          }
        }).catch( (err) => {
          conn.isConnected = false
          errorHandler(res, conn, err)
          res.status(500).json({
            statusCode: 'ERROR',
            message: 'Database connection error.'
          })
          next()
        })
```

**What happened?**
ESLint got whiney about the space before my ES6 function
`.then( (newConn) => {` 
and 
`}).catch( (err) => {`

**What did you expect to happen?**
I would like a way to specify **for arrow functions only** to either require a space after a parenthesis ... ie `.catch(` _or_ to require a space before the variable declaration on an arrow function ... ie `(err) => {`

I find having that space makes it easier to pick out the arrow functions.  I like the standard rule otherwise though, for calling functions `(someVar, anotherVar) =>` and the like, and for `object.function(one, two)`

**Comment by platinumazure** [comment]
Just to generalize a bit, would you be interested in an enhancement where spaces within parentheses are required if the next token is parentheses?

For example:

```js
const result = someFunction( (foo ? "bar" : "baz"), "another arg");
```

And maybe the same with closing parens?

```js
const result = someFunction( (foo ? "bar" : "baz") );
```

Example with a concise arrow function, which could result in parentheses being the penultimate token:

```js
const result = someFunction( (bar) => ({
    foo: bar
}) );
```

**Comment by zeg-io** [comment]
For me personally allowing the arrow function to be the determiner in regard to requiring the space after the parenthesis would be preferable.  

I believe your example has extra unneeded parenthesis for the arrow function

It's fine with me if we had options about the before and after spacing, but I'd only ever use the before..Here's why...

```
const a = someFunction( (bar, fum) => {
    foo: bar
})
const b = someFunction( (bar) => {
    foo: bar
})
const c = someFunction( (fum, bar) => {
    foo: bar
})
const d = object.function(wee, (bar, fum) => ({
    foo: bar,
   giant: fum
})
const r = someFunction(bar).send({
    foo: bar
})
```
Is cleaner to scan and determine which variables are part of the function or not than this:

```
const a = someFunction((bar, fum) => {
    foo: bar
})
const b = someFunction((bar) => {
    foo: bar
})
const c = someFunction((fum, bar) => {
    foo: bar
})
const d = object.function(wee, (bar, fum) => ({
    foo: bar,
   giant: fum
})
const r = someFunction(bar).send({
    foo: bar
})
```

**Comment by platinumazure** [comment]
You're going to want to look at a parsing tool-- your `const a` example isn't parsing how you think it should. The parentheses I was using are not extraneous.

Here's the parse tree I get from your example (look about 2/3 of the way down):

```json
{
  "type": "Program",
  "start": 0,
  "end": 55,
  "body": [
    {
      "type": "VariableDeclaration",
      "start": 0,
      "end": 55,
      "declarations": [
        {
          "type": "VariableDeclarator",
          "start": 6,
          "end": 54,
          "id": {
            "type": "Identifier",
            "start": 6,
            "end": 7,
            "name": "a"
          },
          "init": {
            "type": "CallExpression",
            "start": 10,
            "end": 54,
            "callee": {
              "type": "Identifier",
              "start": 10,
              "end": 22,
              "name": "someFunction"
            },
            "arguments": [
              {
                "type": "ArrowFunctionExpression",
                "start": 23,
                "end": 53,
                "id": null,
                "generator": false,
                "expression": false,
                "params": [
                  {
                    "type": "Identifier",
                    "start": 24,
                    "end": 27,
                    "name": "bar"
                  },
                  {
                    "type": "Identifier",
                    "start": 29,
                    "end": 32,
                    "name": "fum"
                  }
                ],
                "body": {
                  "type": "BlockStatement",
                  "start": 37,
                  "end": 53,
                  "body": [
                    {
                      "type": "LabeledStatement",
                      "start": 43,
                      "end": 51,
                      "body": {
                        "type": "ExpressionStatement",
                        "start": 48,
                        "end": 51,
                        "expression": {
                          "type": "Identifier",
                          "start": 48,
                          "end": 51,
                          "name": "bar"
                        }
                      },
                      "label": {
                        "type": "Identifier",
                        "start": 43,
                        "end": 46,
                        "name": "foo"
                      }
                    }
                  ]
                }
              }
            ]
          }
        }
      ],
      "kind": "const"
    }
  ],
  "sourceType": "module"
}
```

And try this example-- it won't even parse:

```js
const a = someFunction( (bar, fum) => {
    foo: bar,
    foo2: bar2
})
```

So I think you maybe should be thinking about the closing parenthesis.

<hr />

At this point, I am :-1: for making an exception specifically for arrow functions. It seems like a really niche stylistic requirement and I've never seen that required in any popular style guides. Your best bet is to just make a custom rule that does what you want.

**Comment by zeg-io** [comment]
You're right about the bad code.  I was focusing only on the parenthesis around the function and didn't actually try the code.

Below is a correct and compiling function.
```
new sql.connect(conn.connectionString).then( (newConn) => {
  if (newConn.connected) {
    log.good('done.')
    newConn.isConnected = true
    next()
  }
})
```
Also, the error in my prior example was around returning an object (which does require the parens) basically, I wholly didn't put time into reviewing the functionality of the functions themselves and was only looking at the decoration.  My apologies for that.  Clearly if you are returning a JSON object you need to include the parens around it.

I was imagining the innards of the function being multiple lines long, code wise and therefore needing no parens around it (see above).

As for the question of presence in style guides, it is my contention that these style guides are for the most part relying on tools to enforce the styles within, many times it will become more a question of what tools have available rather than setting a standard and hoping tools catch up to your own style guide.  I would imagine only very, very large players could single-handedly push a real change to what is available.

**Comment by kaicataldo** [comment]
I'm closing this issue because it looks like consensus couldn't be reached. While we wish we could accommodate all requests, we have limited resources and do need to prioritize. We've found that issues failing to reach consensus after 21 days tend to never reach consensus, and as a team have decided to close such issues. This doesn't mean the idea isn't interesting or valuable, only that it's not something the team can commit to at the moment.
